feat(core): rename async to waitForAsync to avoid confusing

JiaLiPassion opened the pull request:
@angular/core/testing provide `async` test utility, but the name `async` is
confusing with the javascript keyword `async`. And in some test case, if you
want to use both the `async` from `@angular/core/testing` and `async/await`,
you may have to write the code like this.

```typescript
function asyncFunc(): Promise<{value: number}> {
    return new Promise(res => setTimeout(() => { res({value: 1}); }));
  }

  function timeout(obj) {
    setTimeout(() => {
      obj.value += 1;
    });
  }

  it('test async', async () => {
    const r: {value: number} = await asyncFunc();
    timeout(r);
    setTimeout(() => {
      expect(r.value).toBe(2);
    });
   });
```

So in this PR, the `async` is renamed to `asyncTest`.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?

- [x] Yes
- [ ] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

petebacondarwin reviewed (changes requested):
Reviewed-for: fw-upgrade

Don't forget to update: https://github.com/angular/angular/blob/efdf347f9c04e19c8c4fa8b8c05f4310f42728d5/aio/content/guide/upgrade-setup.md

Otherwise LGTM

And I believe you are also going to rename `fakeAsync()` too.

clydin commented:
Can you provide some details on when a developer would need to use this:
```
it('test async operations', async(async() => {
  const result = await asyncMethod();
  expect(result).toEqual('expected');
}));
```
instead of this:
```
it('test async operations', async () => {
  const result = await asyncMethod();
  expect(result).toEqual('expected');
});
```

jelbourn commented:
@JiaLiPassion I'm personally missing some context on this; is this something that's been previously proposed/discussed with the larger team?

JiaLiPassion commented:
@jelbourn, there is no discussion about this one, I just want to make a proposal by making this PR so it may easier to review and discuss. Thanks.

mary-poppins commented:
You can preview 5a2fb22 at https://pr37583-5a2fb22.ngbuilds.io/.
You can preview bc1464f at https://pr37583-bc1464f.ngbuilds.io/.

petebacondarwin reviewed (approved):
Reviewed-for: fw-upgrade

atscott reviewed (approved):
reviewed for: fw-router

This would need to land in a major release and should have a migration with it, right?

atscott reviewed (approved):
reviewed-for: fw-router

pkozlowski-opensource reviewed (approved):
The change itself LGTM and I see why we would like to rename those functions. If I understand correctly for now we are just deprecating the `async` / `fakeAsync` functions _and_ removing the usage from the main repo? It would have been slightly more clear with 2 separate commits (one adding a new function + depreciation and another one with usage changes).

Having said all this I guess we will have to plan the actual removal very carefully and be sure to have an automatic migration - otherwise it would be too much of the breaking change (as we can see in the main repo).

Reviewed-for: public-api 
Reviewed-for: fw-core 
Reviewed-for: fw-testing

pkozlowski-opensource commented:
Also @JiaLiPassion for some reason CI didn't run all the jobs so I'm adding the cleanup label.

mary-poppins commented:
You can preview 4be26d7 at https://pr37583-4be26d7.ngbuilds.io/.

mary-poppins commented:
You can preview 04c5420 at https://pr37583-04c5420.ngbuilds.io/.

mary-poppins commented:
You can preview 728023e at https://pr37583-728023e.ngbuilds.io/.

JiaLiPassion commented:
@pkozlowski-opensource , thank you for the review, I have rebased and also split the PR into two commits.

> Having said all this I guess we will have to plan the actual removal very carefully and be sure to have an automatic migration - otherwise it would be too much of the breaking change (as we can see in the main repo).

Yeah, I agree, will try to figure out how to make an automatic migration tool later.

mary-poppins commented:
You can preview 4d139ee at https://pr37583-4d139ee.ngbuilds.io/.

mary-poppins commented:
You can preview 092f2a3 at https://pr37583-092f2a3.ngbuilds.io/.

clydin commented:
I believe this change has broader implications than just the FW changes.  For example, tooling will also be affected (both Angular and third-party).  Instructional material such as courseware, videos, blogs/articles, etc. will need to be updated by their providers as well (this can be a time consuming and expensive endeavor).   As mentioned, a migration would also be of importance as developers would receive potentially hundreds of deprecation warnings after upgrading.

I think it may also be useful to have a larger discussion on whether the `async` helper function should continue to exist moving forward.  And if it should not then a renaming may not be necessary (`fakeAsync` is fairly distinctly named in relation to the `async` keyword).

jelbourn commented:
@StephenFluin and @mgechev I'd like to get your input on this change and the naming 

TL;DR: having a function named `async` is problematic when there's a native `async` keyword, so this change proposes deprecating the names `async` and `fakeAsync` in favor of `asyncTest` and `fakeAsyncTest`

alfaproject commented:
I'm with @clydin here. We haven't used `async` ever since we could use native `async/await`. Can't it just be straight up deprecated?

shairez commented:
I get why `async` needs to change (or to get deprecated).

From my experience - I really didn't see a lot of usage for it as well.

But `fakeAsync` is the more commonly used function and I agree with @clydin that this change will have a huge negative impact in multiple aspects.. while not super necessary since there aren't any collisions with native functions. 

So... I say - do what ever you want with `async`, just [leave `fakeAsync` alone](https://www.youtube.com/watch?v=0wG_h3U568w)  😄

yjaaidi commented:
+1 to get rid of `async()` somehow :)
The confusing thing with `asyncTest` is that it sounds weird if used with `beforeEach` or `afterEach` (e.g. `beforeEach(asyncTest(...))` ) as we are not talking about a test here.
Maybe `asyncRun` or `asyncWait`? or the opposite `runAsync` or `waitAsync`?

yjaaidi commented:
By the way, would you agree to replace the component's test schematic:
```typescript
beforeEach(async(() => {
  TestBed.configureTestingModule({})
  .compileComponents();
}));
```
by:
```typescript
beforeEach(async () => {
  await TestBed.configureTestingModule({ })
  .compileComponents();
});
```

this makes tests zoneless friendly by default.


_We might also think about the option below but I am not in favor of it as developers might replace it with a block and forget to add `async / await`._
```typescript
beforeEach(() => TestBed.configureTestingModule({}).compileComponents());
```

cc. @BioPhoton @LayZeeDK

petebacondarwin commented:
Actually I really like this combo: `runAsync` and `fakeAsync` that would resolve the concerns raised by @clydin , @shairez and @yjaaidi.

LayZeeDK commented:
I think both `async` and `asyncTest` are equally poor function names for what this one does.

Like many other Zone.js-related things in Angular, this one is pretty difficult to explain and reason about. If I recall correctly, `async` intercepts all monkey-patched APIs through a zone and flushes them after callback execution.

A more descriptive name might be something like `interceptAndFlushAsync` or `waitForAsync`.

@yjaaidi
Regarding the spec generation schematics, I remember reading somewhere that it's not even necessary to run `TestBed#compileComponents` asynchronously if we're using the Angular CLI.

That it's called the same as a keyword has been an issue since Angular version 2.0 since TypeScript has had `async-await` since version 1.7. I'm with @clydin in that this will deprecate a lot of material including Angular.io's testing guide, Angular's codebase and the codebases of a ton of lot of libraries in the ecosystem.

I would love to get a more descriptive name though.

One other aspect that @yjaaidi brings up is related to the flaky future of Zone.js. Angular appications can't target `es2018` as TypeScript output as long as they're using Zone.js unless we force downleveling of `async-await` into the Angular compiler or add an explicit NgZone trigger after each `async-await` statement (also using the compiler). This would need to affect all code imported from `node_modules` as well.

If removing Zone.js from Angular is within scope of +3 major versions (I don't know the roadmap for Angular+Zone.js), would making a big change like this, preceded by a deprecation period, be worth it, seeing that this is only related to NgZone-enabled applications?

JiaLiPassion commented:
@LayZeeDK, I agree with your points that `async` and `asyncTest` may not be the good names.

> If removing Zone.js from Angular is within scope of +3 major versions (I don't know the roadmap for Angular+Zone.js), would making a big change like this, preceded by a deprecation period, be worth it, seeing that this is only related to NgZone-enabled applications?

`async` and `fakeAsync` are just test helper functions, they have nothing to do with `ngZone`, `async` will call `done` automatically when all `macroTasks and microTasks` finished, and `fakeAsync` just like another version of `jasmine clock and mockDate` to make async test to sync. those functions are using `zone.js` but even Angular doesn't depend on `zone.js` to trigger change detection any longer, the test helper functions still work as expected.

mhevery commented on packages/core/testing/src/async.ts:
```diff
@@ -43,3 +43,12 @@ export function async(fn: Function): (done: any) => any {
 }
+
+/**
+ *
```
Needs docs of how to use. (Even if it says "see ...")

JiaLiPassion replied:
Added JSDoc `@see {@link waitForAsync}`

mhevery commented on packages/core/testing/src/async.ts:
```diff
@@ -43,3 +43,12 @@ export function async(fn: Function): (done: any) => any {
+/**
+ *
+ * @publicApi
+ * @deprecated use asyncTest()
```
```suggestion
 * @deprecated use `asyncTest()` (expected removal in v12)
```

JiaLiPassion replied:
Got it, updated, tahnks.

mhevery commented on packages/core/testing/src/async.ts:
```diff
@@ -25,7 +25,7 @@ import {asyncFallback} from './async_fallback';
  * @publicApi
  */
-export function async(fn: Function): (done: any) => any {
+export function asyncTest(fn: Function): (done: any) => any {
```
`<bikeshedding>`
- `waitForAsync` Describes what the function does. 
- `waitForAsyncToComplete` Describes what the function does in a more complete sentence. 

`</bikeshedding>`

JiaLiPassion replied:
Got it, I will rename it to `waitForAsync`.

mhevery commented on packages/core/testing/src/fake_async.ts:
```diff
@@ -43,14 +43,22 @@ export function resetFakeAsyncZone(): void {
+/**
+ *
+ * @publicApi
+ * @deprecated use fakeAsyncTest()
```
```suggestion
 * @deprecated use `fakeAsyncTest()` (expected removal in v12)
```

mhevery commented on packages/core/testing/src/fake_async.ts:
```diff
@@ -43,14 +43,22 @@ export function resetFakeAsyncZone(): void {
 }
 
+/**
+ *
```
Needs JSDOC

mhevery commented on packages/core/testing/src/fake_async.ts:
```diff
@@ -43,14 +43,22 @@ export function resetFakeAsyncZone(): void {
  * @publicApi
  */
-export function fakeAsync(fn: Function): (...args: any[]) => any {
+export function fakeAsyncTest(fn: Function): (...args: any[]) => any {
```
Not really convinced that renaming `fakeAsync` to `fakeAsyncTest` improves readability.

mhevery commented:
- [ ] The backwards compatible rename can land in patch release. 
- [ ] The automatic migration tool should land before the major release. 
- [ ] Need to discuss with leads if this kind of rename is what we want.

petebacondarwin commented:
I removed the review label and added cleanup because a) it has conflicts and needs a rebase; b) we have agreed (I believe) to use `waitForAsync ` and keep `fakeAsync`.

mary-poppins commented:
You can preview 2982982 at https://pr37583-2982982.ngbuilds.io/.

JiaLiPassion commented:
@petebacondarwin , I have rebased and rename `async` to `waitForAsync`, also revert `fakeAsync`. Please review, thanks.

mary-poppins commented:
You can preview 64d89ad at https://pr37583-64d89ad.ngbuilds.io/.

petebacondarwin commented:
I'm afraid the AIO test code needs reformatting

mary-poppins commented:
You can preview 9c17bc7 at https://pr37583-9c17bc7.ngbuilds.io/.

mary-poppins commented:
You can preview 830ef51 at https://pr37583-830ef51.ngbuilds.io/.

mary-poppins commented:
You can preview 8e8929c at https://pr37583-8e8929c.ngbuilds.io/.

mary-poppins commented:
You can preview f494e53 at https://pr37583-f494e53.ngbuilds.io/.

JiaLiPassion commented:
@petebacondarwin , yeah, I updated the format and commit again, please review, thanks.

petebacondarwin reviewed (approved):
Reviewed-for: fw-upgrade

mary-poppins commented:
You can preview 4a63d18 at https://pr37583-4a63d18.ngbuilds.io/.

mary-poppins commented:
You can preview a0d0dce at https://pr37583-a0d0dce.ngbuilds.io/.

mhevery commented:
[presubmit](http://test/OCL:324084689:BASE:324084717:1596144868193:45b9734f)

mhevery reviewed (approved):
LGTM
- [ ] Sorry to be a pain, but could you break the PR into two. One for the change, and one to move `async` to `waitForAsync` We had it broken up but somehow it gote merged.

Reviewed-for: global-approvers
Reviewed-for: fw-core

JiaLiPassion committed:
feat(core): rename async to waitForAsync to avoid confusing

@angular/core/testing provide `async` test utility, but the name `async` is
confusing with the javascript keyword `async`. And in some test case, if you
want to use both the `async` from `@angular/core/testing` and `async/await`,
you may have to write the code like this.

```typescript
it('test async operations', async(async() => {
  const result = await asyncMethod();
  expect(result).toEqual('expected');
}));
```

So in this PR, the `async` is renamed to `waitForAsync` and also deprecate `async`.

JiaLiPassion committed:
feat(core): update reference and doc to change `async` to `waitAsync`.

The last commit change `async` to `waitForAsync`.
This commit update all usages in the code and also update aio doc.

mary-poppins commented:
You can preview e20abd9 at https://pr37583-e20abd9.ngbuilds.io/.

mhevery commented:
[presubmit](http://test/OCL:324084689:BASE:324276416:1596228236860:99c16274)

mhevery commented:
merge-assistance: global-approval
